src/commands/keygen.rs

```diff
@@ -0,0 +1,380 @@
+            Some(_) => unreachable!(),
+            None => {
+                return Err("Missing algorithm (-a) option".into());
+            }
+        };
+
+        let Some(name) = name else {
+            return Err("Missing domain name argument".into());
+        };
+
+        Ok(Self {
+            algorithm,
+            make_ksk,
+            #[cfg(target_family = "unix")]
+            create_symlinks,
+            #[cfg(target_family = "unix")]
+            force_symlinks,
+            name,
+        })
+    }
+}
+
+impl From<Keygen> for Command {
+    fn from(value: Keygen) -> Self {
+        Self::Keygen(value)
+    }
+}
+
+impl Keygen {
+    fn parse_algorithm(value: &str) -> Result<GenerateParams, clap::Error> {
+        match value {
+            "RSASHA256" => return Ok(GenerateParams::RsaSha256 { bits: 2048 }),
+            "ECDSAP256SHA256" => return Ok(GenerateParams::EcdsaP256Sha256),
+            "ECDSAP384SHA384" => return Ok(GenerateParams::EcdsaP384Sha384),
+            "ED25519" => return Ok(GenerateParams::Ed25519),
+            "ED448" => return Ok(GenerateParams::Ed448),
+            _ => {}
+        }
+
+        // TODO: Remove attrs when more RSA algorithms are added.
+        #[allow(clippy::collapsible_match)]
+        if let Some((name, params)) = value.split_once(':') {
+            #[allow(clippy::single_match)]
+            match name {
+                "RSASHA256" => {
+                    let bits: u32 = params.parse().map_err(|err| {
+                        clap::Error::raw(
+                            clap::error::ErrorKind::InvalidValue,
+                            format!("invalid RSA key size '{params}': {err}"),
+                        )
+                    })?;
+                    return Ok(GenerateParams::RsaSha256 { bits });
+                }
+                _ => {}
+            }
+        }
+
+        Err(clap::Error::raw(
+            clap::error::ErrorKind::InvalidValue,
+            format!("unrecognized algorithm '{value}'"),
+        ))
+    }
+
+    pub fn execute(self, env: impl Env) -> Result<(), Error> {
+        let mut stdout = env.stdout();
+
+        let params = self.algorithm;
+
+        // The digest algorithm is selected based on the key algorithm.
+        let digest_alg = match params.algorithm() {
+            SecAlg::RSASHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ED25519 => DigestAlg::SHA256,
+            SecAlg::ED448 => DigestAlg::SHA256,
+            _ => unreachable!(),
+        };
+
+        // Generate the key.
+        // TODO: Attempt repeated generation to avoid key tag collisions.
+        let (secret_key, public_key) = common::generate(params)
+            .map_err(|err| format!("an implementation error occurred: {err}").into())
+            .context("generating a cryptographic keypair")?;
```

Comment:
We still need this in the other utils. Would be cool if you could look into that at some point.

Reply:
What component are you talking about?  Error context?

Reply:
Yeah, but it's not a suggestion for this PR